cache/disk/disk.go

```diff
@@ -119,32 +135,60 @@ func migrateDirectory(dir string) error {
 // loadExistingFiles lists all files in the cache directory, and adds them to the
 // LRU index so that they can be served. Files are sorted by access time first,
 // so that the eviction behavior is preserved across server restarts.
-func (c *diskCache) loadExistingFiles() error {
-	log.Printf("Loading existing files in %s.\n", c.dir)
-
+func (c *diskCache) loadExistingFiles(dirs []string) error {
 	// Walk the directory tree
 	type NameAndInfo struct {
 		info os.FileInfo
 		name string
 	}
 	var files []NameAndInfo
-	err := filepath.Walk(c.dir, func(name string, info os.FileInfo, err error) error {
-		if !info.IsDir() {
-			files = append(files, NameAndInfo{info, name})
+
+	filesChan := make(chan []NameAndInfo, len(dirs))
+	workChan := make(chan string)
+	workers := runtime.NumCPU()
+
+	for i := 0; i < workers; i++ {
+		go func(workChan <-chan string, filesChan chan<- []NameAndInfo) {
+			for dir := range workChan {
+				dirFiles := make([]NameAndInfo, 0)
+
+				_ = filepath.Walk(dir, func(name string, info os.FileInfo, err error) error {
+					if !info.IsDir() {
+						dirFiles = append(dirFiles, NameAndInfo{info, name})
+					}
+					return nil
+				})
+
+				filesChan <- dirFiles
+			}
+		}(workChan, filesChan)
+	}
+
+	for _, dir := range dirs {
+		workChan <- dir
+	}
+
+	close(workChan)
+
+	i := 0
+	for f := range filesChan {
+		if len(f) > 0 {
+			files = append(files, f...)
+		}
+
+		i++
+		if i == len(dirs) {
+			break
 		}
-		return nil
-	})
-	if err != nil {
-		return err
 	}
 
-	log.Println("Sorting cache files by atime.")
+	close(filesChan)
+
 	// Sort in increasing order of atime
 	sort.Slice(files, func(i int, j int) bool {
 		return atime.Get(files[i].info).Before(atime.Get(files[j].info))
```

Comment:
I guess this means atime.Get is called multiple times for a single file, it is worth calling just once per file and then sorting based on the cached values?

Reply:
(Nah, I don't think so...)